File: #89_Gray_Code/Solution2.java
import java.util.List;
import java.util.ArrayList;

public class Solution2 {
    public List<Integer> grayCode(int n) {
        ArrayList<Integer> list = new ArrayList<Integer>();
        if (n <= 0) {
        	list.add(0);
        	return list;
        }
        list.add(0);
        for (int i = 0; i < n; i++) {
            int nb = 1 << i;
            for (int j = list.size() - 1; j >= 0; j--) {
                list.add(nb + list.get(j));
            }
        }
        return list;
    }

    public static void main(String[] args) {
    	Solution2 sol = new Solution2();
    	List<Integer> rslt = sol.grayCode(3);
    	for (Integer it: rslt) {
    		System.out.print(it + " ");
    	}
    	System.out.println();
    }
}

File: #89_Gray_Code/Solution2.py
class Solution:
	# @param {integer} n
	# @return {integer[]}
	def grayCode(self, n):
		if n == 0:
			return [0]
		if n == 1:
			return [0, 1]
		re = [x for x in range(1 << n)]
		for i in range(1, n):
			for j in range(1 << i, 1 << i + 1):
				re[j] = (1 << i) + re[(1 << i + 1) - 1 - j]
		return re

sol = Solution()
print sol.grayCode(3)